{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### m6A motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### extend to window\n",
    "# m6a_dir = \"/Charles/project/ASm6A/ASm6A/\"\n",
    "# os.chdir(m6a_dir)\n",
    "# #\n",
    "# result_dir = \"motif/ASm6A/window/\"\n",
    "# os.system(\"mkdir -p %s\" % result_dir)\n",
    "# #\n",
    "# bed_list = glob.glob(\"*.bed\")\n",
    "# for bed in bed_list:\n",
    "#     # os.system(\"intersectBed -wo -a %s -b %s | awk -v OFS=\"\\t\" '{print $1,$2,$3,\"*\",\"*\",$10}' | sort -k1,2 | uniq > ${prefix}_bestpeaks.bed\")\n",
    "#     df = pd.read_table(bed, sep=\"\\t\", header=None)\n",
    "#     df.iloc[:,1] = df.iloc[:,1] - 4\n",
    "#     df.iloc[:,2] = df.iloc[:,2] + 4\n",
    "#     res = os.path.join(result_dir, os.path.basename(bed))\n",
    "#     df.to_csv(res, sep=\"\\t\", header=False, index=False)\n",
    "    \n",
    "### get strand info\n",
    "# win_dir = \"/Charles/project/ASm6A/ASm6A/motif/ASm6A/window/\"\n",
    "win_dir = \"/home/galaxy/data/RBP/POSTAR/POSTAR_bed/motif/window/\"\n",
    "os.chdir(win_dir)\n",
    "result_dir = \"../add_strand/\"\n",
    "os.system(\"mkdir -p %s\" % result_dir)\n",
    "gene = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/gencode.v19.annotation_Gene.bed\"\n",
    "#\n",
    "bed_list = glob.glob(\"*.bed\")\n",
    "for bed in bed_list:\n",
    "    res = os.path.join(result_dir, os.path.basename(bed))\n",
    "    os.system(\"intersectBed -a %s -b %s -wa -wb | sort -u > %s\" % (bed, gene, res))\n",
    "\n",
    "### get fasta\n",
    "# stra_dir = \"/Charles/project/ASm6A/ASm6A/motif/ASm6A/add_strand/\"\n",
    "os.chdir(\"../add_strand/\")\n",
    "##\n",
    "! for x in *.bed;do less $x | awk '{OFS=\"\\t\"}{print $1,$2,$3,\".\",\".\",$9}' | sort -k1,1 -k2,2n | uniq > ${x}2;done\n",
    "##\n",
    "result_dir = \"../fasta/\"\n",
    "os.system(\"mkdir -p %s\" % result_dir)\n",
    "fasta = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19.fa\"\n",
    "#\n",
    "bed_list = glob.glob(\"*.bed2\")\n",
    "for bed in bed_list:\n",
    "    res = os.path.join(result_dir, os.path.basename(bed).replace(\".bed2\",'.fa'))\n",
    "    os.system(\"bedtools getfasta -s -fi %s -bed %s -fo %s\" % (fasta, bed, res))\n",
    "    \n",
    "# #### get alt fasta\n",
    "# base_dict = {\"A\": \"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}\n",
    "# fa_dir = \"/Charles/project/ASm6A/ASm6A/motif/ASm6A/fasta/\"\n",
    "# os.chdir(fa_dir)\n",
    "# result_dir = \"../alt_fasta/\"\n",
    "# os.system(\"mkdir -p %s\" % result_dir)\n",
    "# #\n",
    "# raw_dir = \"/Charles/project/ASm6A/ASm6A/motif/ASm6A/add_strand/\"\n",
    "# bed_list = glob.glob(\"%s/*.bed\" % raw_dir)\n",
    "# for bed in bed_list:\n",
    "#     # print(bed)\n",
    "#     prefix, title_dict = os.path.basename(bed).split(\".\")[0], {}\n",
    "#     df = pd.read_table(bed, header=None)\n",
    "#     # >chr10:102116306-102116315(+)    ### 注意修改这一行链的位置信息\n",
    "#     df['title'] = \">\" + df.iloc[:,0] + \":\" + df.iloc[:,1].astype(str) + \"-\" + df.iloc[:,2].astype(str) + \"(\" + df.iloc[:,11] + \")\"\n",
    "#     df[\"alleles\"] = df.iloc[:,3].str.split(\";\").str[0]\n",
    "#     title_dict = dict(zip(df['title'], df['alleles']))\n",
    "#     # print(title_dict['>chr10:101379275-101379284(-)']) \n",
    "#     #\n",
    "#     fasta = os.path.join(fa_dir, prefix+\".fa\")\n",
    "#     fa_dict = {}\n",
    "#     with open(fasta, 'r') as f:\n",
    "#         contents = f.readlines()\n",
    "#         for i in range(len(contents)-1):\n",
    "#             fa_dict[contents[i].strip()] = contents[i+1].strip()\n",
    "#             i += 1\n",
    "#     # print(fa_dict['>chr10:101379275-101379284(-)']) # GTGATTGGA\n",
    "#     #\n",
    "#     result_file = os.path.join(result_dir, os.path.basename(fasta))\n",
    "#     res_list = []\n",
    "#     for title, sequ in fa_dict.items():\n",
    "#         if title in title_dict:\n",
    "#             alleles = title_dict[title]\n",
    "#             if \"(+)\" in title:\n",
    "#                 sequ = list(sequ.upper())\n",
    "#                 assert sequ[4] == alleles.split(\">\")[0]\n",
    "#                 sequ[4] = alleles.split(\">\")[1]\n",
    "#                 res_list.append(title)\n",
    "#                 res_list.append(\"\".join(sequ))\n",
    "#             elif \"(-)\" in title:\n",
    "#                 sequ = list(sequ.upper())\n",
    "#                 assert sequ[4] == base_dict[alleles.split(\">\")[0]]\n",
    "#                 sequ[4] = base_dict[alleles.split(\">\")[1]]\n",
    "#                 res_list.append(title)\n",
    "#                 res_list.append(\"\".join(sequ))\n",
    "#     with open(result_file, 'w') as fw:\n",
    "#         for line in res_list:\n",
    "#             fw.write(line + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identifying m6A motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n",
      "\n",
      "Tool:    bedtools intersect (aka intersectBed)\n",
      "Version: v2.27.1\n",
      "Summary: Report overlaps between two feature files.\n",
      "\n",
      "Usage:   bedtools intersect [OPTIONS] -a <bed/gff/vcf/bam> -b <bed/gff/vcf/bam>\n",
      "\n",
      "\tNote: -b may be followed with multiple databases and/or \n",
      "\twildcard (*) character(s). \n",
      "Options: \n",
      "\t-wa\tWrite the original entry in A for each overlap.\n",
      "\n",
      "\t-wb\tWrite the original entry in B for each overlap.\n",
      "\t\t- Useful for knowing _what_ A overlaps. Restricted by -f and -r.\n",
      "\n",
      "\t-loj\tPerform a \"left outer join\". That is, for each feature in A\n",
      "\t\treport each overlap with B.  If no overlaps are found, \n",
      "\t\treport a NULL feature for B.\n",
      "\n",
      "\t-wo\tWrite the original A and B entries plus the number of base\n",
      "\t\tpairs of overlap between the two features.\n",
      "\t\t- Overlaps restricted by -f and -r.\n",
      "\t\t  Only A features with overlap are reported.\n",
      "\n",
      "\t-wao\tWrite the original A and B entries plus the number of base\n",
      "\t\tpairs of overlap between the two features.\n",
      "\t\t- Overlapping features restricted by -f and -r.\n",
      "\t\t  However, A features w/o overlap are also reported\n",
      "\t\t  with a NULL B feature and overlap = 0.\n",
      "\n",
      "\t-u\tWrite the original A entry _once_ if _any_ overlaps found in B.\n",
      "\t\t- In other words, just report the fact >=1 hit was found.\n",
      "\t\t- Overlaps restricted by -f and -r.\n",
      "\n",
      "\t-c\tFor each entry in A, report the number of overlaps with B.\n",
      "\t\t- Reports 0 for A entries that have no overlap with B.\n",
      "\t\t- Overlaps restricted by -f and -r.\n",
      "\n",
      "\t-v\tOnly report those entries in A that have _no overlaps_ with B.\n",
      "\t\t- Similar to \"grep -v\" (an homage).\n",
      "\n",
      "\t-ubam\tWrite uncompressed BAM output. Default writes compressed BAM.\n",
      "\n",
      "\t-s\tRequire same strandedness.  That is, only report hits in B\n",
      "\t\tthat overlap A on the _same_ strand.\n",
      "\t\t- By default, overlaps are reported without respect to strand.\n",
      "\n",
      "\t-S\tRequire different strandedness.  That is, only report hits in B\n",
      "\t\tthat overlap A on the _opposite_ strand.\n",
      "\t\t- By default, overlaps are reported without respect to strand.\n",
      "\n",
      "\t-f\tMinimum overlap required as a fraction of A.\n",
      "\t\t- Default is 1E-9 (i.e., 1bp).\n",
      "\t\t- FLOAT (e.g. 0.50)\n",
      "\n",
      "\t-F\tMinimum overlap required as a fraction of B.\n",
      "\t\t- Default is 1E-9 (i.e., 1bp).\n",
      "\t\t- FLOAT (e.g. 0.50)\n",
      "\n",
      "\t-r\tRequire that the fraction overlap be reciprocal for A AND B.\n",
      "\t\t- In other words, if -f is 0.90 and -r is used, this requires\n",
      "\t\t  that B overlap 90% of A and A _also_ overlaps 90% of B.\n",
      "\n",
      "\t-e\tRequire that the minimum fraction be satisfied for A OR B.\n",
      "\t\t- In other words, if -e is used with -f 0.90 and -F 0.10 this requires\n",
      "\t\t  that either 90% of A is covered OR 10% of  B is covered.\n",
      "\t\t  Without -e, both fractions would have to be satisfied.\n",
      "\n",
      "\t-split\tTreat \"split\" BAM or BED12 entries as distinct BED intervals.\n",
      "\n",
      "\t-g\tProvide a genome file to enforce consistent chromosome sort order\n",
      "\t\tacross input files. Only applies when used with -sorted option.\n",
      "\n",
      "\t-nonamecheck\tFor sorted data, don't throw an error if the file has different naming conventions\n",
      "\t\t\tfor the same chromosome. ex. \"chr1\" vs \"chr01\".\n",
      "\n",
      "\t-sorted\tUse the \"chromsweep\" algorithm for sorted (-k1,1 -k2,2n) input.\n",
      "\n",
      "\t-names\tWhen using multiple databases, provide an alias for each that\n",
      "\t\twill appear instead of a fileId when also printing the DB record.\n",
      "\n",
      "\t-filenames\tWhen using multiple databases, show each complete filename\n",
      "\t\t\tinstead of a fileId when also printing the DB record.\n",
      "\n",
      "\t-sortout\tWhen using multiple databases, sort the output DB hits\n",
      "\t\t\tfor each record.\n",
      "\n",
      "\t-bed\tIf using BAM input, write output as BED.\n",
      "\n",
      "\t-header\tPrint the header from the A file prior to results.\n",
      "\n",
      "\t-nobuf\tDisable buffered output. Using this option will cause each line\n",
      "\t\tof output to be printed as it is generated, rather than saved\n",
      "\t\tin a buffer. This will make printing large output files \n",
      "\t\tnoticeably slower, but can be useful in conjunction with\n",
      "\t\tother software tools and scripts that need to process one\n",
      "\t\tline of bedtools output at a time.\n",
      "\n",
      "\t-iobuf\tSpecify amount of memory to use for input buffer.\n",
      "\t\tTakes an integer argument. Optional suffixes K/M/G supported.\n",
      "\t\tNote: currently has no effect with compressed files.\n",
      "\n",
      "Notes: \n",
      "\t(1) When a BAM file is used for the A file, the alignment is retained if overlaps exist,\n",
      "\tand excluded if an overlap cannot be found.  If multiple overlaps exist, they are not\n",
      "\treported, as we are only testing for one or more overlaps.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***** ERROR: -b option given, but no database file specified. *****\n",
      "Error: The requested file (.bed) could not be opened. Error message: (No such file or directory). Exiting!\n",
      "\n",
      "\tPosition file = _bestpeaks.bed\n",
      "\tGenome = _home\n",
      "\tOutput Directory = -len\n",
      "!!! 7 not recognized!!\n",
      "\n",
      "\tProgram will find de novo and known motifs in regions in the genome\n",
      "\n",
      "\tUsage: findMotifsGenome.pl <pos file> <genome> <output directory> [additional options]\n",
      "\tExample: findMotifsGenome.pl peaks.txt mm8r peakAnalysis -size 200 -len 8\n",
      "\n",
      "\tPossible Genomes:\n",
      "\t\tmm9\tmouse\n",
      "\t\tdm6\tfly\n",
      "\t\thg38\thuman\n",
      "\t\tmm10\tmouse\n",
      "\t\thg19\thuman\n",
      "\t\t\t-- or --\n",
      "\t\tCustom: provide the path to genome FASTA files (directory or single file)\n",
      "\t\t\tHeads up: will create the directory \"preparsed/\" in same location.\n",
      "\n",
      "\tBasic options:\n",
      "\t\t-mask (mask repeats/lower case sequence, can also add 'r' to genome, i.e. mm9r)\n",
      "\t\t-bg <background position file> (genomic positions to be used as background, default=automatic)\n",
      "\t\t\tremoves background positions overlapping with target positions\n",
      "\t\t\t-chopify (chop up large background regions to the avg size of target regions)\n",
      "\t\t-len <#>[,<#>,<#>...] (motif length, default=8,10,12) [NOTE: values greater 12 may cause the program\n",
      "\t\t\tto run out of memory - in these cases decrease the number of sequences analyzed (-N),\n",
      "\t\t\tor try analyzing shorter sequence regions (i.e. -size 100)]\n",
      "\t\t-size <#> (fragment size to use for motif finding, default=200)\n",
      "\t\t\t-size <#,#> (i.e. -size -100,50 will get sequences from -100 to +50 relative from center)\n",
      "\t\t\t-size given (uses the exact regions you give it)\n",
      "\t\t-S <#> (Number of motifs to optimize, default: 25)\n",
      "\t\t-mis <#> (global optimization: searches for strings with # mismatches, default: 2)\n",
      "\t\t-norevopp (don't search reverse strand for motifs)\n",
      "\t\t-nomotif (don't search for de novo motif enrichment)\n",
      "\t\t-rna (output RNA motif logos and compare to RNA motif database, automatically sets -norevopp)\n",
      "\n",
      "\tScanning sequence for motifs\n",
      "\t\t-find <motif file> (This will cause the program to only scan for motifs)\n",
      "\n",
      "\tKnown Motif Options/Visualization\n",
      "\t\t-mset <vertebrates|insects|worms|plants|yeast|all> (check against motif collects, default: auto)\n",
      "\t\t-basic (just visualize de novo motifs, don't check similarity with known motifs)\n",
      "\t\t-bits (scale sequence logos by information content, default: doesn't scale)\n",
      "\t\t-nocheck (don't search for de novo vs. known motif similarity)\n",
      "\t\t-mcheck <motif file> (known motifs to check against de novo motifs,\n",
      "\t\t-float (allow adjustment of the degeneracy threshold for known motifs to improve p-value[dangerous])\n",
      "\t\t-noknown (don't search for known motif enrichment, default: -known)\n",
      "\t\t-mknown <motif file> (known motifs to check for enrichment,\n",
      "\t\t-nofacts (omit humor)\n",
      "\t\t-seqlogo (use weblogo/seqlogo/ghostscript to generate logos, default uses SVG now)\n",
      "\n",
      "\tSequence normalization options:\n",
      "\t\t-gc (use GC% for sequence content normalization, now the default)\n",
      "\t\t-cpg (use CpG% instead of GC% for sequence content normalization)\n",
      "\t\t-noweight (no CG correction)\n",
      "\t\tAlso -nlen <#>, -olen <#>, see homer2 section below.\n",
      "\n",
      "\tAdvanced options:\n",
      "\t\t-h (use hypergeometric for p-values, binomial is default)\n",
      "\t\t-N <#> (Number of sequences to use for motif finding, default=max(50k, 2x input)\n",
      "\t\t-local <#> (use local background, # of equal size regions around peaks to use i.e. 2)\n",
      "\t\t-redundant <#> (Remove redundant sequences matching greater than # percent, i.e. -redundant 0.5)\n",
      "\t\t-maxN <#> (maximum percentage of N's in sequence to consider for motif finding, default: 0.7)\n",
      "\t\t-maskMotif <motif file1> [motif file 2]... (motifs to mask before motif finding)\n",
      "\t\t-opt <motif file1> [motif file 2]... (motifs to optimize or change length of)\n",
      "\t\t-rand (randomize target and background sequences labels)\n",
      "\t\t-ref <peak file> (use file for target and background - first argument is list of peak ids for targets)\n",
      "\t\t-oligo (perform analysis of individual oligo enrichment)\n",
      "\t\t-dumpFasta (Dump fasta files for target and background sequences for use with other programs)\n",
      "\t\t-preparse (force new background files to be created)\n",
      "\t\t-preparsedDir <directory> (location to search for preparsed file and/or place new files)\n",
      "\t\t-keepFiles (keep temporary files)\n",
      "\t\t-fdr <#> (Calculate empirical FDR for de novo discovery #=number of randomizations)\n",
      "\n",
      "\thomer2 specific options:\n",
      "\t\t-homer2 (use homer2 instead of original homer, default)\n",
      "\t\t-nlen <#> (length of lower-order oligos to normalize in background, default: -nlen 3)\n",
      "\t\t\t-nmax <#> (Max normalization iterations, default: 160)\n",
      "\t\t\t-neutral (weight sequences to neutral frequencies, i.e. 25%, 6.25%, etc.)\n",
      "\t\t-olen <#> (lower-order oligo normalization for oligo table, use if -nlen isn't working well)\n",
      "\t\t-p <#> (Number of processors to use, default: 1)\n",
      "\t\t-e <#> (Maximum expected motif instance per bp in random sequence, default: 0.01)\n",
      "\t\t-cache <#> (size in MB for statistics cache, default: 500)\n",
      "\t\t-quickMask (skip full masking after finding motifs, similar to original homer)\n",
      "\t\t-minlp <#> (stop looking for motifs when seed logp score gets above #, default: -10)\n",
      "\n",
      "\tOriginal homer specific options:\n",
      "\t\t-homer1 (to force the use of the original homer)\n",
      "\t\t-depth [low|med|high|allnight] (time spent on local optimization default: med)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/peak/motif/\")\n",
    "# os.chdir(\"/Charles/project/ASm6A/ASm6A/motif/\")\n",
    "os.chdir(\"/Charles/project/ASm6A/ASm6A/motif/ASm6A/add_strand/total_homer_nt6/\")\n",
    "\n",
    "# bed_file=\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/peak/macs2_Nc1_normalized.bed\"\n",
    "# bed_file = \"m6A.bed\"\n",
    "bed_file = \"/Charles/project/ASm6A/ASm6A/motif/ASm6A/add_strand/total_homer_nt6/ASm6A.bed\"\n",
    "gtf=\"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19_only22.gtf\"\n",
    "fasta=\"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19.fa\"\n",
    "# prefix=\"Nc1\"\n",
    "prefix = \"ASm6A\"\n",
    "!sort -k5,5 -g ${bed_file}| head -10000 | awk '{ print $1\"\\t\"$2\"\\t\"$3}' > ${prefix}.location\n",
    "!intersectBed -wo -a ${prefix}.location -b $gtf | awk -v OFS=\"\\t\" '{print $1,$2,$3,\"*\",\"*\",$10}' | sort -k1,2 | uniq > ${prefix}_bestpeaks.bed\n",
    "!bedtools getfasta -s -fi $fasta -bed ${prefix}_bestpeaks.bed -fo ${prefix}_bestpeaks.fa\n",
    "# ame -oc ${prefix}_ame ${prefix}_bestpeaks.fa $RRACH_motif\n",
    "!findMotifsGenome.pl ${prefix}_bestpeaks.bed $fasta ${prefix}_homer -len 6 -rna -p 15 # ${THREAD_NUM:=1}\n",
    "#/home/software/meme-5.0.2/scripts/dreme -k 7 -oc ${prefix}_dreme -p ${prefix}_bestpeaks.fa -rna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. RRACH sequence within peak¶\n",
    "#### 2. motif enrichment in ASm6A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:40: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "388426 333885 722311\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "\n",
    "os.chdir(\"/Charles/project/ASm6A/ASm6A/motif/m6A/RRACH/\")\n",
    "gene = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/gencode.v19.annotation_Gene.bed\"\n",
    "m6a = \"../m6A.bed\"\n",
    "res = \"Allsamples_hg19.overlapgene\"\n",
    "os.system(\"intersectBed -a %s -b %s -wa -wb | sort -u > %s\" % (m6a, gene, res))\n",
    "! less Allsamples_hg19.overlapgene | awk '{OFS=\"\\t\"}{print $1,$2,$3,\".\",\".\",$9}' | sort -k1,2 | uniq > Allsamples_hg19.overlapgene2\n",
    "fasta = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19.fa\"\n",
    "bed = res + \"2\"\n",
    "res = \"Allsamples_hg19.fa\"\n",
    "os.system(\"bedtools getfasta -s -fi %s -bed %s -fo %s\" % (fasta, bed, res))\n",
    "\n",
    "motif = \"/home/galaxy/project/alleleSpecific_analysis/data/motif/m6A_motif.homer\"\n",
    "fa = res\n",
    "res = \"Allsamples_hg19.txt\"\n",
    "os.system(\"findMotifs.pl %s fasta homer/ -find %s > %s\" % (fa, motif, res))\n",
    "\n",
    "txt = \"Allsamples_hg19.txt\"\n",
    "result_file = \"Allsamples_hg19.sites\"\n",
    "df = pd.read_table(txt, sep=\"\\t\")\n",
    "df['chr'] = df['FASTA ID'].str.split(\":\").str[0]\n",
    "df['strand'] = df['FASTA ID'].str.split(\"(\").str[1].str.split(\")\").str[0]\n",
    "df['start'] = df['FASTA ID'].str.split(\":\").str[1].str.split(\"-\").str[0].astype(int)\n",
    "df['end'] = df['FASTA ID'].str.split(\":\").str[1].str.split(\"-\").str[1].str.split(\"(\").str[0].astype(int)\n",
    "###\n",
    "df_same = df[df['Sequence'] == df['Motif Name']]\n",
    "df_p = df_same[df_same['strand'] == \"+\"]\n",
    "# df_p['site'] = (df_p['end'] - 4) + df_p['Offset'] +2\n",
    "df_p['site'] = (df_p['start']+1+df_p['end'])/2 + df_p['Offset'] +2\n",
    "#\n",
    "df_n = df_same[df_same['strand'] == \"-\"]\n",
    "# df_n['site'] = (df_n['end'] - 4) - df_n['Offset'] -2\n",
    "df_n['site'] = (df_n['start']+1+df_n['end'])/2 - df_n['Offset'] -2\n",
    "df_same = pd.concat([df_p, df_n])\n",
    "###\n",
    "df_reverse = df[df['Sequence'] != df['Motif Name']]\n",
    "df_p = df_reverse[df_reverse['strand'] == \"+\"]\n",
    "# df_p['site'] = (df_p['end'] - 4) + df_p['Offset'] -2\n",
    "df_p['site'] = (df_p['start']+1+df_p['end'])/2 + df_p['Offset'] -2\n",
    "#\n",
    "df_n = df_reverse[df_reverse['strand'] == \"-\"]\n",
    "# df_n['site'] = (df_n['end'] - 4) - df_n['Offset'] +2\n",
    "df_n['site'] = (df_n['start']+1+df_n['end'])/2 - df_n['Offset'] +2\n",
    "df_reverse = pd.concat([df_p, df_n])\n",
    "\n",
    "df = pd.concat([df_same, df_reverse])\n",
    "# print(len(df_same), len(df_reverse), len(df))\n",
    "site_list = df['site'].tolist()\n",
    "site_list = [math.ceil(x) for x in site_list]\n",
    "df['site'] = site_list\n",
    "# df['s'] = df['site'] - 1\n",
    "df['start'] = df['site'] - 3\n",
    "df['end'] = df['site'] + 2\n",
    "# df[['chr','s','site','FASTA ID']].to_csv(result_file, sep=\"\\t\", header=False, index=False)\n",
    "df[['chr','start','end','FASTA ID']].to_csv(result_file, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "\n",
    "### filter by letter: a,A,T,t\n",
    "! bedtools getfasta -fi /home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19.fa -bed Allsamples_hg19.sites > Allsamples_hg19.sites.fa\n",
    "# letter_list = [\"A\", \"a\", \"T\", \"t\"]\n",
    "motif_list = ['AAACC', 'GGTCT', 'AGTCC', 'AGTTC', 'TGTTC', 'TGTTT', 'AAACT', 'GGACC', 'AGTTT', 'AGACT','GAACA',\n",
    " 'AGACC', 'GAACC', 'TGTCC', 'GGACT', 'GAACT', 'AGACA', 'AAACA', 'AGTCT', 'GGTTT', 'GGACA', 'GGTCC', 'TGTCT',\n",
    " 'GGTTC']\n",
    "result_list = []\n",
    "with open(\"Allsamples_hg19.sites.fa\", 'r') as f:\n",
    "    contents = f.readlines()\n",
    "    for i in range(len(contents)-1):\n",
    "        if contents[i+1].strip().upper() in motif_list:\n",
    "            result_list.append(contents[i])\n",
    "            \n",
    "with open(\"RRACH.bed\", 'w') as fw:\n",
    "    for x in result_list:\n",
    "        chro = x.split(\">\")[1].split(\":\")[0]\n",
    "        if (chro != \"chrX\") and (chro != \"chrY\"):\n",
    "            start = x.split(\":\")[1].split(\"-\")[0]\n",
    "            end = x.split(\"-\")[1]\n",
    "            fw.write(\"%s\\t%s\\t%s\" % (chro, start, end))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
