{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### select the SNP list which appeared in both conditions\n",
    "### must satisfy the minimum read number\n",
    "\n",
    "def read_file(in_file):\n",
    "    df = pd.read_table(in_file, sep=\"\\t\")\n",
    "    df['term'] = df['contig'] + \":\" + df['position'].astype(str)\n",
    "    return df['term'].tolist()\n",
    "\n",
    "def select_baseline_SNPs(in_dir):\n",
    "    os.chdir(in_dir)\n",
    "#     f_1, f_2 = \"WT.txt\", \"KD.txt\"\n",
    "    f_1, f_2 = \"WT.txt\", \"D395A.txt\"\n",
    "    t1_list = read_file(f_1)\n",
    "    t2_list = read_file(f_2)\n",
    "    SNV_list = set(t1_list).intersection(set(t2_list))\n",
    "    return SNV_list\n",
    "\n",
    "# data_dir = \"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/Hypothesis/Fisher/heter_sites/\"\n",
    "# asm6A_list = select_baseline_SNPs(data_dir)\n",
    "asm6A_list = select_baseline_SNPs(data_dir)\n",
    "\n",
    "print(len(asm6A_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### filter > IP/INPUT >= 1,2 or 3;\n",
    "# os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/mouse_ASm6A/count/Hypothesis/comparision/ASm6A/\")\n",
    "# # os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/Hypothesis/comparision/ASm6A/\")\n",
    "\n",
    "# def filter_file(in_file):\n",
    "#     df = pd.read_table(in_file)\n",
    "#     df.columns = ['chr','s','e','info','ref','alt']\n",
    "#     df['mark'] = df.iloc[:,3].str.split(\";\").str[2]\n",
    "#     ## IP/Input > 2\n",
    "#     df_res = df[((df['mark']==\"ref\") & (df['ref']>=2)) | ((df['mark']==\"alt\") & (df['alt']>=2))]\n",
    "#     del df_res['mark']\n",
    "#     return df_res\n",
    "\n",
    "# result_dir = \"filter/\"\n",
    "# os.system(\"mkdir -p %s\" % result_dir)\n",
    "\n",
    "# # nc = \"WT.bed\"\n",
    "# # kd = \"KD.bed\"\n",
    "# nc = \"WT.bed\"\n",
    "# kd = \"D395A.bed\"\n",
    "# df_nc, df_kd = filter_file(nc), filter_file(kd)\n",
    "# df_nc.to_csv(\"filter/%s\"%nc, sep=\"\\t\", index=False, header=False)\n",
    "# df_kd.to_csv(\"filter/%s\"%kd, sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################### different filter methods ####################################\n",
    "######### filter: within m6A peaks\n",
    "wt_peak = \"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/peak/rm_m6Am/WT_m6A.bed\"\n",
    "kd_peak = \"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/peak/rm_m6Am/KD_m6A.bed\"\n",
    "\n",
    "####### filter: select high \n",
    "result_dir = \"highFC/\"\n",
    "os.system(\"mkdir -p %s\" % result_dir)\n",
    "\n",
    "def filter_file(in_file):\n",
    "    res = \"%s/%s\"% (result_dir, in_file)\n",
    "    df = pd.read_table(in_file)\n",
    "    df = df[((df.iloc[:,5]/df.iloc[:,4])>2) | ((df.iloc[:,5]/df.iloc[:,4])<0.5)]\n",
    "    df.to_csv(\"highFC/%s\"%in_file, sep=\"\\t\", index=False, header=False)\n",
    "#     return df\n",
    "# filter_file(\"Nc1.bed\")\n",
    "# filter_file(\"Fus2.bed\")\n",
    "filter_file(\"WT.bed\")\n",
    "filter_file(\"D395A.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### stat ASm6A change\n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/Hypothesis/comparision/ASm6A/filter/withinm6A/\")\n",
    "nc = \"WT.bed\"\n",
    "kd = \"KD.bed\"\n",
    "\n",
    "def get_termlist(in_file):\n",
    "    df = pd.read_table(in_file)\n",
    "    df['term'] = df.iloc[:,0] + \":\" + df.iloc[:,2].astype(str)\n",
    "    return df['term'].tolist()\n",
    "\n",
    "nc_list, kd_list = get_termlist(nc), get_termlist(kd)\n",
    "print(\"seperately: \", len(nc_list), len(kd_list))\n",
    "nc_list = set(nc_list).intersection(set(asm6A_list))\n",
    "kd_list = set(kd_list).intersection(set(asm6A_list))\n",
    "print(\"based on common heter: \", len(nc_list), len(kd_list))\n",
    "print(\"#########################################\")\n",
    "yes_to_no = set(nc_list).difference(set(kd_list))\n",
    "print(\"yes_to_no: %d\" % len(yes_to_no))\n",
    "with open(\"yes_to_no.txt\", 'w') as fw:\n",
    "    fw.writelines([\"%s\\n\"%x for x in list(yes_to_no)])\n",
    "###\n",
    "no_to_yes = set(kd_list).difference(set(nc_list))\n",
    "print(\"no_to_yes: %d\" % len(no_to_yes))\n",
    "###\n",
    "still_yes = set(nc_list).intersection(set(kd_list))\n",
    "print(\"still yes: %d\" % len(still_yes))\n",
    "\n",
    "### stat the proportion of no_to_yes to most-condifent ASm6A sites\n",
    "# # base_dir = \"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/read_matrix/most_confident/\"\n",
    "# # wt_list = pd.read_table(\"%s/Nc1_highConf.txt\"%base_dir)['ID'].tolist()\n",
    "# # kd_list = pd.read_table(\"%s/Fus2_highConf.txt\"%base_dir)['ID'].tolist()\n",
    "# #\n",
    "# high_yes_to_no = set(yes_to_no).intersection(set(wt_list))\n",
    "# high_prop = len(high_yes_to_no)/len(yes_to_no)\n",
    "# print(len(high_yes_to_no), high_prop)\n",
    "# #\n",
    "# high_no_to_yes = set(no_to_yes).intersection(set(kd_list))\n",
    "# high_prop = len(high_no_to_yes)/len(no_to_yes)\n",
    "# print(len(high_no_to_yes), high_prop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# still yes\n",
    "nc = \"WT.bed\"\n",
    "kd = \"KD.bed\"\n",
    "\n",
    "def read_bed(in_file, data_type):\n",
    "    df = pd.read_table(in_file)\n",
    "    df['AR_%s'%data_type] = df.iloc[:,3].str.split(\";\").str[1].astype(float)\n",
    "    df['term'] = df.iloc[:,0] + \":\" + df.iloc[:,2].astype(str)\n",
    "    return df[['term', 'AR_%s'%data_type]]\n",
    "    \n",
    "def extract_ar():\n",
    "    df_nc, df_kd = read_bed(nc, \"nc\"), read_bed(kd, \"kd\")\n",
    "    df_nc = df_nc[df_nc['term'].isin(still_yes)]\n",
    "    df_kd = df_kd[df_kd['term'].isin(still_yes)]\n",
    "    df_t = df_nc.merge(df_kd, on=\"term\")\n",
    "    return df_t\n",
    "\n",
    "df = extract_ar()\n",
    "df['diff'] = df['AR_nc'] - df['AR_kd']\n",
    "a = df[(df['AR_nc']>0.5) & (df['AR_kd']<0.5)]\n",
    "b = df[(df['AR_nc']<0.5) & (df['AR_kd']>0.5)]\n",
    "print(len(a), len(b), len(still_yes)-len(a)-len(b))\n",
    "#\n",
    "c = df[(df['AR_nc']>0.5) & (df['AR_kd']>0.5)]\n",
    "d = df[(df['AR_nc']<0.5) & (df['AR_kd']<0.5)]\n",
    "cd = pd.concat([c,d])\n",
    "cd.to_csv(\"unchanged_ASm6A.bed\", sep=\"\\t\", index=False)\n",
    "# print(len(df[df['diff']>0]), len(df[df['diff']<=0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## merge wt and kd\n",
    "\n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/new_count/sig/Fisher/\")\n",
    "wt, kd = \"Nc1.txt\", \"Fus2.txt\"\n",
    "\n",
    "def read_file(in_file, data_type):\n",
    "    df = pd.read_table(in_file)\n",
    "    df = df[[\"contig\",\"position\",\"refRPKM_ratio\", \"altRPKM_ratio\", \"allelicRatio\"]]\n",
    "    df.columns = [\"contig\",\"position\",\"refRPKM_%s\"%data_type, \"altRPKM_%s\"%data_type, \"AR_%s\"%data_type]\n",
    "    return df\n",
    "df_wt, df_kd = read_file(wt, \"wt\"), read_file(kd, \"kd\")\n",
    "df = df_wt.merge(df_kd, on=[\"contig\",\"position\"])\n",
    "df['term'] = df.iloc[:,0] + \":\" + df.iloc[:,1].astype(str)\n",
    "df = df.round(2)\n",
    "res = \"total.counts\"\n",
    "df.to_csv(res, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### \n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/new_count/sig/Fisher/\")\n",
    "total = \"total.counts\"\n",
    "df = pd.read_table(total)\n",
    "df_res = df[df['term'].isin(no_to_yes)]\n",
    "res = \"no_to_yes.txt\"\n",
    "# df_res = df[df['term'].isin(yes_to_no)]\n",
    "# res = \"yes_to_no.txt\"\n",
    "df_res.to_csv(res, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### compare the read count\n",
    "# os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/sig/comparision/ASE/\")\n",
    "# ase1 = \"Nc1.bed\"\n",
    "# ase2 = \"Fus2.bed\"\n",
    "# os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/\")\n",
    "# txt_list = glob.glob(\"*.txt\")\n",
    "# #\n",
    "# name_dict = {\"Fus2-inp_L4_710D01.readcounts.txt\":\"Fus2-input\", \"Fus2-ip_L4_712D01.readcounts.txt\":\"Fus2-ip\", \"Nc1-inp_L4_709D01.readcounts.txt\":\"Nc1-input\", \"Nc1-ip_L4_711D01.readcounts.txt\":\"Nc1-ip\"}\n",
    "# ###\n",
    "# df_total = pd.DataFrame()\n",
    "# ###\n",
    "# df = pd.read_table(txt_list[0])\n",
    "# df = df[['contig','position','refCount','altCount']]\n",
    "# p = name_dict[txt_list[0]]\n",
    "# df.columns = ['contig','position','refCount_%s'%p,'altCount_%s'%p]\n",
    "# df_total = df\n",
    "# ##################\n",
    "# for txt in txt_list[1:]:\n",
    "#     print(txt)\n",
    "#     df = pd.read_table(txt)\n",
    "#     df = df[['contig','position','refCount','altCount']]\n",
    "#     p = name_dict[txt]\n",
    "#     df.columns = ['contig','position','refCount_%s'%p,'altCount_%s'%p]\n",
    "#     df_total = df_total.merge(df)\n",
    "# df_total['term'] = df_total['contig'] + \":\" + df_total['position'].astype(str)\n",
    "# df_total.to_csv(\"total.readcounts.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_file(in_file):\n",
    "    df = pd.read_table(in_file, sep=\"\\t\")\n",
    "    ### qvalue < 0.05\n",
    "    df = df[df['qvalue'] < 0.05]\n",
    "    ###############################\n",
    "    df['term'] = df['contig'] + \":\" + df['position'].astype(str)\n",
    "    term_list = df['term'].tolist()\n",
    "    # term_list = [x for x in term_list if x in SNV_list]\n",
    "    #####\n",
    "    df['mark'] = np.where(df['allelicRatio']>0.5, \"alt\", \"ref\")\n",
    "    mark_list = df['mark'].tolist()\n",
    "    return term_list, mark_list\n",
    "\n",
    "def stat_number():\n",
    "    t1_list, m1_list = read_file(f_1)\n",
    "    t2_list, m2_list = read_file(f_2)\n",
    "    print(len(set(t1_list)), len(set(t2_list)))\n",
    "    yes_to_no = set(t1_list).difference(set(t2_list))\n",
    "    yes_still_yes = set(t1_list).intersection(set(t2_list))\n",
    "    no_to_yes = set(t2_list).difference(set(t1_list))\n",
    "    print(len(yes_to_no), len(yes_still_yes), len(no_to_yes))\n",
    "    return list(set(t1_list)), list(set(yes_to_no)), list(set(yes_still_yes)), list(set(no_to_yes))\n",
    "    \n",
    "### ASm6A change\n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/sig/Fisher/\")\n",
    "f_1, f_2 = \"Nc1.txt\", \"Fus2.txt\"\n",
    "m6a_0, m6a_1, m6a_2, m6a_3 = stat_number()\n",
    "print(\"###################################\")\n",
    "#### ASE change\n",
    "# os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/mouse_ASm6A/MeRIP/sig/binomi/\")\n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/sig/binomi/\")\n",
    "f_1, f_2 = \"Nc1-inp.txt\", \"Fus2-inp.txt\"\n",
    "ase_0, ase_1, ase_2, ase_3 = stat_number()\n",
    "print(\"###################################\")\n",
    "overlap_0 = set(m6a_0).intersection(set(ase_0))\n",
    "m6a_loss = [x for x in m6a_1 if x in overlap_0]\n",
    "m6a_ase_loss = set(m6a_loss).intersection(set(ase_1))\n",
    "print(len(overlap_0), len(m6a_loss), len(m6a_ase_loss))\n",
    "#m6a_ase_overlap = set(m6a_1).intersection(set(ase_1))\n",
    "#print(len(overlap_0), len(m6a_ase_overlap))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### pick out oppo direction\n",
    "contig_list = [x.split(\":\")[0] for x in m6a_ase_loss]\n",
    "position = [int(x.split(\":\")[1]) for x in m6a_ase_loss]\n",
    "df_target = pd.DataFrame({\"contig\": contig_list, \"position\": position})\n",
    "# print(df_target.head())\n",
    "###\n",
    "asm6a = \"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/sig/Fisher/Nc1.txt\"\n",
    "ase = \"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/sig/binomi/Nc1-inp.txt\"\n",
    "\n",
    "def overlap_and_mark(in_file, data_type):\n",
    "    df = pd.read_table(in_file, sep=\"\\t\")\n",
    "    df = df[df['qvalue'] < 0.05]\n",
    "    df = df.merge(df_target, on=['contig', 'position'], how=\"right\").dropna()\n",
    "    df['mark'] = np.where(df['allelicRatio']>0.5, \"alt\", \"ref\")\n",
    "    df['start'] = df['position'] - 1\n",
    "    df = df[['contig', 'start', 'position', 'refAllele', 'altAllele', 'mark']]\n",
    "    df.columns = ['contig', 'start', 'position', 'refAllele', 'altAllele', 'mark_%s'%data_type]\n",
    "    return df\n",
    "\n",
    "df_asm6a = overlap_and_mark(asm6a, \"asm6a\")\n",
    "df_ase = overlap_and_mark(ase, \"ase\")\n",
    "#\n",
    "df_raw = df_asm6a.merge(df_ase, on=['contig', 'start', 'position']).dropna()\n",
    "print(len(df_raw))\n",
    "# print(df_raw.head())\n",
    "df_res = df_raw[df_raw['mark_asm6a'] != df_raw['mark_ase']]\n",
    "print(len(df_res))\n",
    "res = \"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/sig/oppo_direction/results.bed2\"\n",
    "# df_res.to_csv(res, sep=\"\\t\", index=False)  # header=False, "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trending larger or smaller"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   chr10  122668295  122668296  G>A;0.08;alt  0.48   5.68\n",
      "0  chr10   46957455   46957456  G>A;0.42;alt  5.11   7.04\n",
      "1  chr10   64564891   64564892  G>T;0.29;alt  1.01   2.42\n",
      "2   chr1  107600578  107600579  C>T;0.58;ref  7.32   5.31\n",
      "3   chr1  110753764  110753765  G>C;0.26;alt  4.30  12.51\n",
      "4  chr11   10800531   10800532  G>A;0.36;alt  3.32   5.98\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr10</th>\n",
       "      <th>6275222</th>\n",
       "      <th>6275223</th>\n",
       "      <th>G&gt;A;0.26;alt</th>\n",
       "      <th>2.75</th>\n",
       "      <th>7.93</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>63958111</td>\n",
       "      <td>63958112</td>\n",
       "      <td>T&gt;C;0.03;alt</td>\n",
       "      <td>0.46</td>\n",
       "      <td>13.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>64966931</td>\n",
       "      <td>64966932</td>\n",
       "      <td>G&gt;A;1.0;ref</td>\n",
       "      <td>11.81</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109637228</td>\n",
       "      <td>109637229</td>\n",
       "      <td>C&gt;G;0.66;ref</td>\n",
       "      <td>9.06</td>\n",
       "      <td>4.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>108380381</td>\n",
       "      <td>108380382</td>\n",
       "      <td>G&gt;A;1.0;ref</td>\n",
       "      <td>18.66</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>108384206</td>\n",
       "      <td>108384207</td>\n",
       "      <td>C&gt;T;0.78;ref</td>\n",
       "      <td>5.64</td>\n",
       "      <td>1.56</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr10    6275222    6275223  G>A;0.26;alt   2.75   7.93\n",
       "0  chr10   63958111   63958112  T>C;0.03;alt   0.46  13.31\n",
       "1  chr10   64966931   64966932   G>A;1.0;ref  11.81   0.00\n",
       "2   chr1  109637228  109637229  C>G;0.66;ref   9.06   4.77\n",
       "3  chr11  108380381  108380382   G>A;1.0;ref  18.66   0.00\n",
       "4  chr11  108384206  108384207  C>T;0.78;ref   5.64   1.56"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/data/shFUS/count/Hypothesis/comparision/ASm6A/filter/withinm6A/\")\n",
    "kd, wt = \"KD.bed\", \"WT.bed\"\n",
    "df_kd, df_wt = pd.read_table(kd, header=None), pd.read_table(wt, header=None)\n",
    "print(df_kd.head())\n",
    "df_wt.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
