{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### generate allele count script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### index vcf\n",
    "# java -jar /home/cjr/software/gatk-4.1.2.0/gatk-package-4.1.2.0-local.jar IndexFeatureFile\n",
    "# -F ind1_gatk_het_SNP.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 90 samples\n",
    "ind_dict = {\n",
    "    \"ind1\": ['brain_IP_1', 'heart_IP_1', 'liver_IP_1'],\n",
    "    \"ind2\":\n",
    "    ['brain_IP_2', 'heart_IP_2', 'placenta_IP_2', 'kidney_IP_2', 'liver_IP_2'],\n",
    "    \"ind3\": ['brain_IP_3', 'heart_IP_3', 'liver_IP_3', 'kidney_IP_3'],\n",
    "    \"ind4\": [\n",
    "        'stomach_IP_4', 'muscle_IP_4', 'lung_IP_4', 'kidney_IP_4',\n",
    "        'placenta_IP_4'\n",
    "    ],\n",
    "    \"ind5\": ['stomach_IP_5', 'muscle_IP_5', 'lung_IP_5'],\n",
    "    \"ind6\": ['placenta_IP_6'],\n",
    "    \"ind7\": [\n",
    "        'Adipose-1-1', 'Adrenal_gland-1-1', 'Aorta-1-1', 'Heart-1-1',\n",
    "        'Skin-1-1', 'Spleen-1-1'\n",
    "    ],\n",
    "    \"ind8\": [\n",
    "        'Lung-2-1', 'Lung-2-4', 'Spleen-2-1', 'Tongue-2-1',\n",
    "        'Urinary_bladder-2-1'\n",
    "    ],\n",
    "    \"ind9\":\n",
    "    ['Appendix-3-2', 'Colon-3-2', 'Esophagus-3-2', 'Muscle-3-2', 'Spleen-3-2'],\n",
    "    \"ind10\": [\n",
    "        'Aorta-4-2', 'Esophagus-4-2', 'Heart-4-2', 'Jejunum-4-2', 'Liver-4-2',\n",
    "        'Lung-4-2', 'Lung-4-4', 'Prostate-4-2', 'Rectum-4-2', 'Skin-4-2',\n",
    "        'Stomach-4-2', 'Testis-4-2', 'Thyroid_gland-4-2', 'Urinary_bladder-4-2'\n",
    "    ],\n",
    "    \"ind11\": [\n",
    "        'Appendix-5-3', 'Brainstem-5-3', 'Cerebellum-5-3', 'Cerebrum-5-3',\n",
    "        'Duodenum-5-3', 'Hypothalamus-5-3', 'Jejunum-5-3', 'Muscle-5-3',\n",
    "        'Rectum-5-3', 'Stomach-5-3', 'Thyroid_gland-5-3', 'Trachea-5-3',\n",
    "        'Urinary_bladder-5-3'\n",
    "    ],\n",
    "    \"ind12\": ['Cerebrum-6-3'],\n",
    "    \"ind13\": ['Cerebellum-7-4'],\n",
    "    \"ind14\": [\n",
    "        'FrontalCortex_1', 'Cerebellum_1', 'Heart_1', 'Liver_1', 'Lung_1',\n",
    "        'Kidney_1', 'Spleen_1'\n",
    "    ],\n",
    "    \"ind15\": [\n",
    "        'FrontalCortex_2', 'Cerebellum_2', 'Heart_2', 'Liver_2', 'Kidney_2',\n",
    "        'Spleen_2', 'Muscle_1'\n",
    "    ],\n",
    "    \"ind16\": [\n",
    "        'FrontalCortex_3', 'Cerebellum_3', 'Heart_3', 'Liver_3', 'Lung_2',\n",
    "        'Muscle_2'\n",
    "    ],\n",
    "    \"ind17\": ['Muscle_3'],\n",
    "    \"ind18\": ['Lung_3', 'Kidney_3', 'Spleen_3']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# os.chdir(\"/Charles/project/ASm6A/by_ind/count/\")\n",
    "os.chdir(\n",
    "    \"/home/galaxy/project/alleleSpecific_analysis/data/bam_wasp/corrected_count/\"\n",
    ")\n",
    "\n",
    "vcf_dir = \"/Charles/project/ASm6A/by_ind/SNP_calling/filtered_vcf/\"\n",
    "# input_bam_dir = \"/Charles/project/ASm6A/corrected_input/\"\n",
    "# ip_bam_dir = \"/Charles/project/ASm6A/corrected_ip/\"\n",
    "input_bam_dir = \"/home/galaxy/project/alleleSpecific_analysis/data/bam_wasp/corrected_input/\"\n",
    "ip_bam_dir = \"/home/galaxy/project/alleleSpecific_analysis/data/bam_wasp/corrected_ip/\"\n",
    "\n",
    "prefix = \"python /home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/run_GATK_ASEReadCounter.py /usr/local/bin/GenomeAnalysisTK.jar /home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19.fa\"\n",
    "# postfix = \"--disable_drf\"\n",
    "\n",
    "\n",
    "def alter_ip_name():\n",
    "    map_file = \"/home/galaxy/project/alleleSpecific_analysis/data/sample_info/map.txt\"\n",
    "    df = pd.read_table(map_file, header=None)\n",
    "    map_dict = dict(zip(df.iloc[:, 1], df.iloc[:, 0]))\n",
    "    return map_dict\n",
    "\n",
    "\n",
    "def get_input_name():\n",
    "    map_file = \"/home/galaxy/project/alleleSpecific_analysis/data/sample_info/total_Input_vs_IP.txt\"\n",
    "    df = pd.read_table(map_file)\n",
    "    sample_dict = dict(zip(df['IP'], df['Input']))\n",
    "    return sample_dict\n",
    "\n",
    "\n",
    "map_dict = alter_ip_name()\n",
    "sample_dict = get_input_name()\n",
    "with open(\"run.sh\", 'w') as fw:\n",
    "    for ind, ip_list in ind_dict.items():\n",
    "        vcf = os.path.join(vcf_dir, \"%s_gatk_het_SNP.vcf\" % ind)\n",
    "        for ip_sample in ip_list:\n",
    "            input_sample, ip_bam = \"\", \"\"\n",
    "            ip_bam = os.path.join(ip_bam_dir,\n",
    "                                  \"%s/%s-uniq.bam\" % (ip_sample, ip_sample))\n",
    "            if not os.path.exists(ip_bam):\n",
    "                ip_bam = os.path.join(\n",
    "                    ip_bam_dir, \"%s/%s-uniq.bam\" %\n",
    "                    (map_dict[ip_sample], map_dict[ip_sample]))\n",
    "            if ip_sample in sample_dict:\n",
    "                input_sample = sample_dict[ip_sample]\n",
    "            else:\n",
    "                input_sample = sample_dict[map_dict[ip_sample]]\n",
    "            input_bam = os.path.join(\n",
    "                input_bam_dir, \"%s/%s-uniq.bam\" % (input_sample, input_sample))\n",
    "            #\n",
    "            ip_result_dir = \"ip_count/%s\" % os.path.basename(ip_bam).split(\n",
    "                \"-uniq.bam\")[0]\n",
    "            input_result_dir = \"input_count/%s\" % os.path.basename(\n",
    "                input_bam).split(\"-uniq.bam\")[0]\n",
    "            #\n",
    "            fw.write(\"%s %s %s %s\\n\" %\n",
    "                     (prefix, vcf, ip_bam, ip_result_dir))  # , postfix\n",
    "            fw.write(\"%s %s %s %s\\n\" %\n",
    "                     (prefix, vcf, input_bam, input_result_dir))  # , postfix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### nohup less run.sh | parallel -j20 &"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
