{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "mobile-plasma",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### select ASm6A sites for validation\n",
    "#### There are two independent data sets\n",
    "\n",
    "os.chdir(\"/Charles/project/ASm6A/data/validation_ASm6A/LBFC20201232/count/\")\n",
    "in_file = \"ASm6As.bed\"\n",
    "prefix = \"ASm6As\"\n",
    "##\n",
    "# os.chdir(\"/Charles/project/ASm6A/ASm6A/corrRBP/motifbreakR/FUS_related_ASm6As/\")\n",
    "# in_file = \"totalSamples_positive.bed\"\n",
    "# prefix = \"totalSamples-ASm6As\"\n",
    "###\n",
    "flank_bed = \"%s_flank.bed\" % prefix\n",
    "# result_overlapFUS_bed = \"%s_flank_FUS.bed\" % prefix\n",
    "# result_overlapFUS_overlapPeak_bed = \"%s_flank_FUS_Peak.bed\" % prefix\n",
    "# final_result = \"%s_finalResults.bed\" % prefix\n",
    "\n",
    "fus = \"/Charles/project/ASm6A/data/RBP/union/union.bed\"\n",
    "#     fus = \"/Charles/project/ASm6A/data/RBP/union/ContainingFUS_genes.bed\"\n",
    "m6a_peak = \"/Charles/project/ASm6A/shFUS/mapping/human/peak/merged_peak_MSPC/wt1/ConsensusPeaks.bed\"\n",
    "#     m6a_peak = \"/Charles/project/ASm6A/shFUS/mapping/human/peak/merged_peak_MSPC/normalized_macs/wt1.bed\"\n",
    "\n",
    "def extract_flankRegion(in_file):\n",
    "    df = pd.read_table(in_file, header=None)\n",
    "    df['chr'] = df.iloc[:, 0]\n",
    "    df['start'] = df.iloc[:, 1] - 100\n",
    "    df['end'] = df.iloc[:, 2] + 100\n",
    "    df['name'] = df.iloc[:, 3]\n",
    "    df[['chr', 'start', 'end', 'name']].drop_duplicates().to_csv(result_bed,\n",
    "                                                         sep=\"\\t\",\n",
    "                                                         header=False,\n",
    "                                                         index=False)\n",
    "\n",
    "def overlap_bed(bed1, bed2, result_bed):\n",
    "    os.system(\"bedtools intersect -a %s -b %s -wa -wb > %s\" %\n",
    "              (bed1, bed2, result_bed))\n",
    "# def overlap_FUS():\n",
    "#     os.system(\"bedtools intersect -a %s -b %s -wa -wb > %s\" %\n",
    "#               (result_bed, fus, result_overlapFUS_bed))\n",
    "\n",
    "# def overlap_293T():\n",
    "#     os.system(\n",
    "#         \"bedtools intersect -a %s -b %s -wa -wb > %s\" %\n",
    "#         (result_overlapFUS_bed, m6a_peak, result_overlapFUS_overlapPeak_bed))\n",
    "\n",
    "def calc_dist(in_file, out_file):\n",
    "    with open(out_file, 'w') as fw:\n",
    "        with open(in_file, 'r') as f:\n",
    "            for line in f.readlines():\n",
    "                ASm6A_position = int(line.split(\"\\t\")[2]) - 100 ####\n",
    "                FUS_start = int(line.split(\"\\t\")[6])\n",
    "                FUS_end = int(line.split(\"\\t\")[7])\n",
    "                v_list = [int(x) for x in range(FUS_start+1, FUS_end+1)]\n",
    "                distance = min([abs(x-ASm6A_position) for x in v_list])\n",
    "                fw.write(line.strip() + \"\\t%s\\n\" % distance)\n",
    "    \n",
    "# def extract_fq():\n",
    "#     genome = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19.fa\"\n",
    "#     result_fa = result_overlapFUS_overlapPeak_bed.replace(\".bed\", \".fa\")\n",
    "#     os.system(\"bedtools getfasta -fi %s -bed %s -name+ > %s\" %\n",
    "#               (genome, result_overlapFUS_overlapPeak_bed, result_fa))\n",
    "def extract_fq(bed, result_file):\n",
    "    genome = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19.fa\"\n",
    "    result_fa = os.path.basename(bed).replace(\".bed\", \".fa\")\n",
    "    os.system(\"bedtools getfasta -fi %s -bed %s -name+ > %s\" %\n",
    "              (genome, bed, result_fa))\n",
    "    with open(result_file, 'w') as fw:\n",
    "        with open(result_fa, 'r') as f:\n",
    "            contents = f.readlines()\n",
    "            for i in range(len(contents)-1):\n",
    "                if contents[i].startswith(\">\"):\n",
    "                    title = contents[i].strip()\n",
    "                    seq = contents[i+1].strip()\n",
    "                    chro = title.split(\"::\")[1].split(\":\")[0]\n",
    "                    start = title.split(\":\")[-1].split(\"-\")[0]\n",
    "                    end = title.split(\"-\")[-1]\n",
    "                    name = title.split(\">\",1)[1].split(\"::\")[0]\n",
    "                    new_line = \"\"\n",
    "                    if name != \"\":\n",
    "                        new_line = \"%s\\t%s\\t%s\\t%s\\t%s\" % (chro, start, end, name, seq)\n",
    "                    else:\n",
    "                        new_line = \"%s\\t%s\\t%s\\t%s\" % (chro, start, end, seq)\n",
    "                    fw.write(new_line + \"\\n\")\n",
    "    os.remove(result_fa)\n",
    "\n",
    "\n",
    "extract_flankRegion(in_file)\n",
    "extract_fq(flank_bed, \"flank_withSeq.bed\")\n",
    "extract_fq(fus, \"fus_withSeq.bed\")\n",
    "extract_fq(m6a_peak, \"m6aPeak_withSeq.bed\")\n",
    "\n",
    "overlap_bed(\"flank_withSeq.bed\", \"fus_withSeq.bed\", \"fus_asm6a_withSeq.bed\")\n",
    "overlap_bed(\"fus_asm6a_withSeq.bed\", \"m6aPeak_withSeq.bed\", \"fus_asm6a_m6aPeak_withSeq.bed\")\n",
    "calc_dist(\"fus_asm6a_m6aPeak_withSeq.bed\", \"fus_asm6a_m6aPeak_results.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "improved-bennett",
   "metadata": {},
   "source": [
    "#### pick out ASm6As from 90 tissue samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "expensive-porcelain",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Charles/project/ASm6A/ASm6A/FUS_validate/\")\n",
    "####\n",
    "fus = \"/Charles/project/ASm6A/data/RBP/union/union.bed\"\n",
    "#     fus = \"/Charles/project/ASm6A/data/RBP/union/ContainingFUS_genes.bed\"\n",
    "asm6a = \"total.bed\"\n",
    "m6a_peak = \"/Charles/project/ASm6A/shFUS/mapping/human/peak/merged_peak_MSPC/wt1/ConsensusPeaks.bed\"\n",
    "\n",
    "def overlap_bed(bed1, bed2, result_bed):\n",
    "    os.system(\"bedtools intersect -a %s -b %s -wa -wb > %s\" %\n",
    "              (bed1, bed2, result_bed))\n",
    "    \n",
    "def extract_fq(bed, result_file):\n",
    "    genome = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19.fa\"\n",
    "    result_fa = os.path.basename(bed).replace(\".bed\", \".fa\")\n",
    "    os.system(\"bedtools getfasta -fi %s -bed %s -name+ > %s\" %\n",
    "              (genome, bed, result_fa))\n",
    "    with open(result_file, 'w') as fw:\n",
    "        with open(result_fa, 'r') as f:\n",
    "            contents = f.readlines()\n",
    "            for i in range(len(contents)-1):\n",
    "                if contents[i].startswith(\">\"):\n",
    "                    title = contents[i].strip()\n",
    "                    seq = contents[i+1].strip()\n",
    "                    chro = title.split(\"::\")[1].split(\":\")[0]\n",
    "                    start = title.split(\":\")[-1].split(\"-\")[0]\n",
    "                    end = title.split(\"-\")[-1]\n",
    "                    new_line = \"%s\\t%s\\t%s\\t%s\" % (chro, start, end, seq)\n",
    "                    fw.write(new_line + \"\\n\")\n",
    "#                     i += 1\n",
    "    os.remove(result_fa)\n",
    "                \n",
    "extract_fq(fus, \"fus_withSeq.bed\")\n",
    "# extract_fq(asm6a, \"asm6a_withSeq.bed\")\n",
    "extract_fq(m6a_peak, \"m6aPeak_withSeq.bed\")\n",
    "\n",
    "overlap_bed(\"fus_withSeq.bed\", asm6a, \"fus_asm6a_withSeq.bed\")\n",
    "overlap_bed(\"fus_asm6a_withSeq.bed\", \"m6aPeak_withSeq.bed\", \"fus_asm6a_m6aPeak_withSeq.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "derived-secret",
   "metadata": {},
   "outputs": [],
   "source": [
    "### select motif\n",
    "in_file = \"fus_asm6a_m6aPeak_withSeq.bed\"\n",
    "with open(\"fus_asm6a_m6aPeak_withMotif.bed\", 'w') as fw:\n",
    "    with open(in_file, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            fus_seq = line.split(\"\\t\")[3]\n",
    "            if (\"GGTG\" in fus_seq):\n",
    "                asm6a_pos = int(line.split(\"\\t\")[6])\n",
    "                fus_start = int(line.split(\"\\t\")[1])\n",
    "                motif_index = fus_seq.index(\"GGTG\")\n",
    "                start = fus_start + motif_index + 1\n",
    "                end = start + 3\n",
    "                dist = min(abs(start-asm6a_pos), abs(end-asm6a_pos))\n",
    "                fw.write(line.strip() + \"\\t%s\\n\" % str(dist))\n",
    "            if (\"CCAC\" in fus_seq):\n",
    "                asm6a_pos = int(line.split(\"\\t\")[6])\n",
    "                fus_start = int(line.split(\"\\t\")[1])\n",
    "                motif_index = fus_seq.index(\"CCAC\")\n",
    "                start = fus_start + motif_index + 1\n",
    "                end = start + 3\n",
    "                dist = min(abs(start-asm6a_pos), abs(end-asm6a_pos))\n",
    "                fw.write(line.strip() + \"\\t%s\\n\" % str(dist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "endless-gateway",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "æ— ",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
